packages/medusa/src/utils/transaction/transaction-builder.ts

```diff
@@ -0,0 +1,358 @@
+        )
+        if (index > -1 && actionStep.next) {
+          if (actionStep.next) {
+            parentStep.next[index] = actionStep.next as InternalStep
+          } else {
+            parentStep.next.splice(index, 1)
+          }
+        }
+      } else {
+        parentStep.next = actionStep.next
+      }
+      this.updateDepths(
+        actionStep.next as InternalStep,
+        parentStep,
+        1,
+        parentStep.depth
+      )
+    }
+    return this
+  }
+
+  pruneAction(action: string) {
+    const parentStep = this.findParentStepByAction(action, this.steps)
+    if (parentStep) {
+      const actionStep = this.findStepByAction(action)
+      if (actionStep) {
+        if (Array.isArray(parentStep.next)) {
+          const index = parentStep.next.findIndex(
+            (step) => step.action === action
+          )
+          if (index > -1) {
+            parentStep.next.splice(index, 1)
+          }
+        } else {
+          delete parentStep.next
+        }
+      }
+    }
+    return this
+  }
+
+  private findStepByAction(
+    action: string,
+    step: InternalStep = this.steps
+  ): InternalStep | undefined {
+    if (step.action === action) {
+      return step
+    }
+
+    if (Array.isArray(step.next)) {
+      for (const subStep of step.next) {
+        const found = this.findStepByAction(action, subStep as InternalStep)
+        if (found) {
+          return found
+        }
+      }
+    } else if (step.next && typeof step.next === "object") {
+      return this.findStepByAction(action, step.next as InternalStep)
+    }
+
+    return
+  }
+
+  private findParentStepByAction(
+    action: string,
+    step: InternalStep = this.steps
+  ): InternalStep | undefined {
+    if (!step.next) {
+      return
+    }
+
+    const nextSteps = Array.isArray(step.next) ? step.next : [step.next]
+    for (const nextStep of nextSteps) {
+      if (!nextStep) {
+        continue
+      }
+      if (nextStep.action === action) {
+        return step
+      }
+      const foundStep = this.findParentStepByAction(
+        action,
+        nextStep as InternalStep
+      )
+      if (foundStep) {
+        return foundStep
+      }
+    }
+
+    return
+  }
+
+  private updateDepths(
+    startingStep: InternalStep,
+    parent,
+    incr = 1,
+    beginFrom?: number
+  ): void {
+    if (!startingStep) {
+      return
+    }
+
+    const update = (step: InternalStep, parent, beginFrom) => {
+      step.depth = beginFrom + incr
+      step.parent = parent.action
+      if (Array.isArray(step.next)) {
+        step.next.forEach((nextAction) => update(nextAction, step, step.depth))
+      } else if (step.next) {
+        update(step.next, step, step.depth)
+      }
+    }
+    update(startingStep, parent, beginFrom ?? startingStep.depth)
+  }
+
+  build(): TransactionStepsDefinition {
+    if (!this.steps.next) {
+      return {}
+    }
+
+    const ignore = ["depth", "parent"]
+    const result = JSON.parse(
```

Comment:
question: why parse/stringify?

Reply:
to cleanup the properties `depth` and `parent` that are not being used at the moment, but I do see utility to it in the future.

Reply:
got it, make sense 👍